core/state/statedb_hooked.go

```diff
@@ -0,0 +1,242 @@
+	s.inner.AddAddressToAccessList(addr)
+}
+
+func (s *hookedStateDB) AddSlotToAccessList(addr common.Address, slot common.Hash) {
+	s.inner.AddSlotToAccessList(addr, slot)
+}
+
+func (s *hookedStateDB) PointCache() *utils.PointCache {
+	return s.inner.PointCache()
+}
+
+func (s *hookedStateDB) Prepare(rules params.Rules, sender, coinbase common.Address, dest *common.Address, precompiles []common.Address, txAccesses types.AccessList) {
+	s.inner.Prepare(rules, sender, coinbase, dest, precompiles, txAccesses)
+}
+
+func (s *hookedStateDB) RevertToSnapshot(i int) {
+	s.inner.RevertToSnapshot(i)
+}
+
+func (s *hookedStateDB) Snapshot() int {
+	return s.inner.Snapshot()
+}
+
+func (s *hookedStateDB) AddPreimage(hash common.Hash, bytes []byte) {
+	s.inner.Snapshot()
+}
+
+func (s *hookedStateDB) Witness() *stateless.Witness {
+	return s.inner.Witness()
+}
+
+func (s *hookedStateDB) SubBalance(addr common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) uint256.Int {
+	prev := s.inner.SubBalance(addr, amount, reason)
+	if s.hooks.OnBalanceChange != nil && !amount.IsZero() {
+		newBalance := new(uint256.Int).Sub(&prev, amount)
+		s.hooks.OnBalanceChange(addr, prev.ToBig(), newBalance.ToBig(), reason)
+	}
+	return prev
+}
+
+func (s *hookedStateDB) AddBalance(addr common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) uint256.Int {
+	prev := s.inner.AddBalance(addr, amount, reason)
+	if s.hooks.OnBalanceChange != nil && !amount.IsZero() {
+		newBalance := new(uint256.Int).Add(&prev, amount)
+		s.hooks.OnBalanceChange(addr, prev.ToBig(), newBalance.ToBig(), reason)
+	}
+	return prev
+}
+
+func (s *hookedStateDB) SetNonce(address common.Address, nonce uint64) {
+	s.inner.SetNonce(address, nonce)
+	if s.hooks.OnNonceChange != nil {
+		s.hooks.OnNonceChange(address, nonce-1, nonce)
+	}
+}
+
+func (s *hookedStateDB) SetCode(address common.Address, code []byte) {
+	s.inner.SetCode(address, code)
+	if s.hooks.OnCodeChange != nil {
+		s.hooks.OnCodeChange(address, types.EmptyCodeHash, nil, crypto.Keccak256Hash(code), code)
+	}
+}
+
+func (s *hookedStateDB) SetState(address common.Address, key common.Hash, value common.Hash) common.Hash {
+	prev := s.inner.SetState(address, key, value)
+	if s.hooks.OnStorageChange != nil && prev != value {
+		s.hooks.OnStorageChange(address, key, prev, value)
+	}
+	return prev
+}
+
+func (s *hookedStateDB) SelfDestruct(address common.Address) uint256.Int {
+	prev := s.inner.SelfDestruct(address)
+	if !prev.IsZero() {
+		if s.hooks.OnBalanceChange != nil {
+			s.hooks.OnBalanceChange(address, prev.ToBig(), new(big.Int), tracing.BalanceDecreaseSelfdestruct)
+		}
+	}
+	return prev
+}
+
+func (s *hookedStateDB) Selfdestruct6780(address common.Address) (uint256.Int, bool) {
+	prev, changed := s.inner.Selfdestruct6780(address)
+	if !prev.IsZero() && changed {
+		if s.hooks.OnBalanceChange != nil {
+			s.hooks.OnBalanceChange(address, prev.ToBig(), new(big.Int), tracing.BalanceDecreaseSelfdestruct)
```

Comment:
I suspect this is now wrong. This balancechange here only triggers for the `decrease`, so presumably theres a corresponding `increase` trigger somewhere. It may be that `selfdestruct(self)` now triggers the `+` but not the `-`. I'll check it once I have a functional IDE again.. Would be good to have this covered by a testcase

Reply:
So, this happens prior
```go
func opSelfdestruct6780(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
	if interpreter.readOnly {
		return nil, ErrWriteProtection
	}
	beneficiary := scope.Stack.pop()
	balance := interpreter.evm.StateDB.GetBalance(scope.Contract.Address())
	interpreter.evm.StateDB.SubBalance(scope.Contract.Address(), balance, tracing.BalanceDecreaseSelfdestruct)
	interpreter.evm.StateDB.AddBalance(beneficiary.Bytes20(), balance, tracing.BalanceIncreaseSelfdestruct)
	interpreter.evm.StateDB.Selfdestruct6780(scope.Contract.Address())
```
Regardless of who the recipient is (self or other), and whether the contract is new or old, both `decrese` and `increase` fires here. 

The cases that exist are

1. `new` contract, beneficiary `other`: 
   1. decrease-event in opcode
   2. increase-event in opcode
   3. No events triggered in statedb. 
2. `new` contract, beneficiary `self`:
   1. decrease-event in opcode
   2. increase-event in opcode
   3. triggers decrease-event in statedb + sets to zero
3. `old` contract, beneficiary `other`:
   1. decrease-event in opcode
   2. increase-event in opcode
   3. No events triggered in statedb.
4. `old` contract, beneficiary `self`:
   1. decrease-event in opcode
   2. increase-event in opcode
   3. No events triggered in statedb.


I think all of these cases are handled correctly.

Reply:
Actually, except for one thing. The `2.iii`, when a new contract is self-destructed with self-recipient, then it should be traced as a `burn`, not a `balancedecreasselfdestruct` (beneficiary). 

There are basically two burns: one immediate when a contract does selfdestruct-to-self, and one later on, which only ever happens if someone sends some funds to an already destructed account. 

So as I see it, line `204` and `214` should use `BalanceDecreaseSelfdestructBurn`. 

However, it seems @s1na has already considered this and decided to not handle it that way: 
```
	// BalanceDecreaseSelfdestructBurn is ether that is sent to an already self-destructed
	// account within the same tx (captured at end of tx).
	// Note it doesn't account for a self-destruct which appoints itself as recipient.
	BalanceDecreaseSelfdestructBurn BalanceChangeReason = 14
```
So, in other words, all good!